File: src/main/java/me/rsman/BetterMinecraftCore/Enchantments/Protection.java
package me.rsman.BetterMinecraftCore.Enchantments;

import me.rsman.BetterMinecraftCore.Managers.ItemTypeChecker;
import org.bukkit.NamespacedKey;
import org.bukkit.enchantments.Enchantment;
import org.bukkit.enchantments.EnchantmentTarget;
import org.bukkit.inventory.ItemStack;

import java.util.HashMap;
import java.util.Map;

public class Protection extends CustomEnchantClass {
    private static Protection instance;

    private static final Map<String, Long> attributeModifiers = new HashMap<String, Long>(){{
        put("defense", 5L);
        put("health", 15L);
    }};

    public Protection() {
        super(NamespacedKey.minecraft("bmc_protection"));
    }

    @Override
    public String getName() {
        return "bmc_Protection";
    }

    @Override
    public int getMaxLevel() {
        return 5;
    }

    @Override
    public int getStartLevel() {
        return 1;
    }

    @Override
    public EnchantmentTarget getItemTarget() {
        return EnchantmentTarget.ARMOR;
    }

    @Override
    public boolean isTreasure() {
        return false;
    }

    @Override
    public boolean isCursed() {
        return false;
    }

    @Override
    public boolean conflictsWith(Enchantment enchantment) {
        return false;
    }

    @Override
    public boolean canEnchantItem(ItemStack itemStack) {
        return true;
    }

    public static Protection getEnchant(){
        if (instance == null) {
            instance = new Protection();
        }
        return instance;
    }

    @Override
    public boolean isApplicable(ItemStack item) {
        return ItemTypeChecker.isArmorOrHead(item);
    }

    @Override
    public int getMinimumLevel() {
        return 99;
    }

    @Override
    public boolean hasAttributesModifiers() {
        return false;
    }

    @Override
    public Map<String, Long> getAttributesModifiers() {
        return Protection.attributeModifiers;
    }
}

File: src/main/kotlin/me/rsman/BetterMinecraftCore/Enchantments/Protection.kt
package me.rsman.BetterMinecraftCore.Enchantments

import java.util.HashMap
import org.bukkit.enchantments.Enchantment
import org.bukkit.NamespacedKey
import org.bukkit.inventory.ItemStack
import me.rsman.BetterMinecraftCore.Managers.ItemTypeChecker
import me.rsman.BetterMinecraftCore.enums.EAttributes
import org.bukkit.enchantments.EnchantmentTarget

class Protection : CustomEnchantClass(NamespacedKey.minecraft("bmc_protection")) {

    override fun getName(): String {
        return "bmc_Protection"
    }

    override fun getMaxLevel(): Int {
        return 5
    }

    override fun getStartLevel(): Int {
        return 1
    }

    override fun getItemTarget(): EnchantmentTarget {
        return EnchantmentTarget.ARMOR
    }

    override fun isTreasure(): Boolean {
        return false
    }

    override fun isCursed(): Boolean {
        return false
    }

    override fun conflictsWith(enchantment: Enchantment): Boolean {
        return false
    }

    override fun canEnchantItem(itemStack: ItemStack): Boolean {
        return true
    }

    override fun isApplicable(item: ItemStack): Boolean {
        return ItemTypeChecker.isArmorOrHead(item)
    }

    override val minimumLevel: Int
        get() = 99

    override fun hasAttributesModifiers(): Boolean {
        return attributeModifiers.isNotEmpty()
    }

    override val attributeModifiers: Map<EAttributes, Long>
        get() = _attributeModifiers

    companion object {
        private lateinit var _instance: Protection
        private var _attributeModifiers: Map<EAttributes, Long> = mapOf(
            EAttributes.DEFENSE to 5L,
            EAttributes.HEALTH to 15L
        )
        val enchant: Protection
            get() = if(this::_instance.isInitialized) _instance else {
                _instance = Protection(); _instance
            }
    }
}